examples/app_gradio.py

```diff
@@ -0,0 +1,106 @@
+import PIL.Image
+from vqasynth.datasets import Dataloader
+from vqasynth.embeddings import EmbeddingGenerator
+from PIL import Image
+import os
+from vqasynth.embeddings import TagFilter
+from vqasynth.depth import DepthEstimator
+from vqasynth.localize import Localizer
+from vqasynth.scene_fusion import SpatialSceneConstructor
+from vqasynth.prompts import PromptGenerator
+from datasets import DatasetDict, Dataset
+import tempfile
+
+import open3d as o3d
+import os
+import gradio as gr
+import PIL
+
+cache_dir = './vqasynth_output'
+images = 'images'
+
```

Comment:
for the point clouds, this one does shows 3d visuals:

`def merge_pcd_to_obj(pcd_files, output_obj_file, depth_map, rgb_image, focal_length):
    from PIL import Image
    import numpy as np
    import open3d as o3d

    # Resize depth map to match RGB image dimensions if needed
    if depth_map.size != rgb_image.size:
        depth_map = depth_map.resize(rgb_image.size, Image.NEAREST)

    # Convert depth and RGB images to Open3D format
    depth_array = np.array(depth_map).astype(np.float32)
    rgb_array = np.array(rgb_image)
    
    # Ensure depth array is scaled correctly
    depth_array = depth_array / np.max(depth_array) * 1000  # Adjust scaling if necessary

    depth_o3d = o3d.geometry.Image(depth_array)
    rgb_o3d = o3d.geometry.Image(rgb_array)
    rgbd_image = o3d.geometry.RGBDImage.create_from_color_and_depth(
        rgb_o3d, depth_o3d, depth_scale=1000.0, depth_trunc=3.0, convert_rgb_to_intensity=False
    )

    # Define camera intrinsics
    intrinsic = o3d.camera.PinholeCameraIntrinsic()
    intrinsic.set_intrinsics(
        width=rgb_image.width,
        height=rgb_image.height,
        fx=focal_length,
        fy=focal_length,
        cx=rgb_image.width / 2,
        cy=rgb_image.height / 2
    )

    # Create point cloud from the RGBD image
    point_cloud = o3d.geometry.PointCloud.create_from_rgbd_image(rgbd_image, intrinsic)

    # Apply statistical outlier removal
    cl, ind = point_cloud.remove_statistical_outlier(nb_neighbors=20, std_ratio=2.0)
    point_cloud = point_cloud.select_by_index(ind)

    ply_file = "./tmp/output.ply"
    o3d.io.write_point_cloud(ply_file, point_cloud)

    mesh = o3d.io.read_triangle_mesh(ply_file)
    
    o3d.io.write_triangle_mesh(output_obj_file, mesh)

    print(f"Merged point cloud saved as mesh to {output_obj_file}")`

Reply:
But what about `pcd_files`? This is the list of segmented pointclouds representing the identified objects. It looks like you're making a pointcloud from the full scene but is there a way to create a single `.ply` file from the segments put together?

Reply:
yes that's I figuring it out, I tried creating with saved pcd files but they are showing just single lines, maybe I need to debug from inside where pcd is generating and saving.

Reply:
> I tried creating with saved pcd files but they are showing just single lines
Are you using the depth map when you create the pointclouds from the pcd files? This may be the cause

Reply:
I tried debugging the SpatialSceneConstructor, here is the possible fix of the run function which was causing some issues, here is the code which is working:

`def run(self, image_filename, image, depth_map, focallength, masks, output_dir):
        img_array = np.array(image)
        depth_array = np.array(depth_map)

        width, height = image.size

        cx, cy = width / 2, height / 2

        points = []
        colors = []
        height, width = depth_array.shape

        point_cloud_data = []

        for idx, mask in enumerate(masks):
            mask_array = np.array(mask)
            for v in range(height):
                for u in range(width):
                    if mask_array[v, u] > 0:  
                        z = depth_array[v, u] / 1000.0  
                        if z == 0:
                            continue
                        x = (u - cx) * z / focallength
                        y = (v - cy) * z / focallength
                        points.append([x, y, z])

                        if img_array is not None:
                            color = img_array[v, u] / 255.0 
                            colors.append(color)
            

            point_cloud = o3d.geometry.PointCloud()
            point_cloud.points = o3d.utility.Vector3dVector(np.array(points))

            
            if colors:
                point_cloud.colors = o3d.utility.Vector3dVector(np.array(colors))

            
            output_pointcloud_dir = os.path.join(output_dir, 'pointclouds')
            pointcloud_filepath = os.path.join(
                    output_pointcloud_dir,
                    f"pointcloud_{Path(image_filename).stem}_{idx}.ply"
                )
            self.save_pointcloud(point_cloud, pointcloud_filepath)
            point_cloud_data.append(pointcloud_filepath)

        return point_cloud_data, False`